plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+{- | Float bindings inwards. A binding can be floated if it is a lazy binding, or
+ a strict binding whose RHS is a value. Doing so does not have much impact on the
+ program size since it simply moves bindings around without changing the number
+ of bindings, but it may lead to more efficient evaluation.
+
+ Unlike the float-in in GHC, we /can/ float bindings into lambda abstractions,
+ because doing so cannot duplicate work. Therefore we simply float a binding
+ inwards to the deepest possible position.
+
+ We avoid moving a strict binding whose RHS is not a value, because doing so
+ may either change the semantics (if the RHS has effects) or duplicate work
+ (if it is floated into a lambda abstraction).
+
+ Floating inwards is done in a single pass. The subterms are processed bottom-up,
+ during which we keep track of the set of variable @Unique@s within each subterm.
+ The time complexity is linear wrt AST size (strictly speaking it is /O(n*log n)/
+ where the /log n/ comes from `Set` operations).
+
+ Currently we only move non-recursive term bindings. In the future we should also
+ implement float-in for other kinds of bindings.
+-}
+module PlutusIR.Transform.LetFloatIn (floatTerm) where
+
+import PlutusCore qualified as PLC
+import PlutusCore.Builtin qualified as PLC
+import PlutusCore.Name qualified as PLC
+import PlutusIR
+import PlutusIR.Purity (isPure)
+import PlutusIR.Transform.Rename ()
+
+import Control.Lens hiding (Strict)
+import Control.Monad ((<=<))
+import Data.Set (Set)
+import Data.Set qualified as Set
+
+type Uniques = Set PLC.TermUnique
+
+-- | Float bindings in the given `Term` inwards.
+floatTerm ::
+    forall m tyname name uni fun a.
+    ( PLC.HasUnique name PLC.TermUnique
+    , PLC.HasUnique tyname PLC.TypeUnique
+    , PLC.ToBuiltinMeaning uni fun
+    , PLC.MonadQuote m
+    ) =>
+    PLC.BuiltinVersion fun ->
+    Term tyname name uni fun a ->
+    m (Term tyname name uni fun a)
+floatTerm ver = pure . fmap fst . go <=< PLC.rename
+  where
+    -- | Float bindings in the given `Term` inwards, and calculate the set of
+    -- variable `Unique`s in the result `Term`.
+    go ::
+        Term tyname name uni fun a ->
+        Term tyname name uni fun (a, Uniques)
```

Comment:
I guess I don't have to use annotation. I did it just because `Uniques` conceptually _is_ a kind of annotation - we are annotating the Terms and Bindings with `Uniques` during the process.

I think using `(Uniques, Term tyname name uni fun a)` would be more or less equivalent. I don't see how it simplifies anything. `Writer`, on the other hand, would be inconvenient to use, because I'd be calling `runWriter` all the time.

Reply:
It means you're not constantly fetching things in and out of the annotation. You'd just be matching on a simple tuple returned from `go` instead of then also having to mangle the annotations back and forth constantly. I'd have to try it to be sure.

Reply:
Actually I think using annotation is much more convenient. Because in `Term (ann, Uniques)`, each subterm is also annotated with `Uniques`, which is what's needed, since I need to keep descending. `(Term ann, Uniques)` won't work.

Reply:
Yes, I figured out that you do genuinely need it for later processing. I still think it might make the annotating-part simpler but whatever.